File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) return;

        //Initialize fast & slow pointers
        ListNode slow = head;
        ListNode fast = head;

        ListNode prev = null;
        ListNode nxt = null;

        //Reverse the linked list
        while(fast != null && fast.next != null){
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        prev.next = null;
        prev = null;

        while(slow != null){
            nxt = slow.next;
            slow.next = prev;
            prev = slow;
            slow = nxt;
        }

        ListNode start = head;
        ListNode end = prev;
        ListNode startNxt = null;
        ListNode endNxt = null;

        // Check if reversed properly
        // while(ptr2 != null){
        //     System.out.println(ptr2.val);
        //     ptr2 = ptr2.next;
        // }

        // Reorder the list now
        head = start;
        while(start != null && end != null){
            startNxt = start.next;
            endNxt = end.next;

            start.next = end;
            if (startNxt == null) break;
            end.next = startNxt;

            start = startNxt;
            end = endNxt;
        }
    }
}

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        
        # D, 1, 2, 3, 4, 5

        if left == right:
            return head

        dummy = ListNode()
        dummy.next = head

        prev = dummy
        curr = dummy.next

        for _ in range(left-1):
            prev = curr
            curr = curr.next
        
        prev.next = None

        p = None
        n = curr
        for _ in range(right-left+1):
            nxt = curr.next
            curr.next = p
            p = curr
            curr = nxt
        
        n.next = curr
        prev.next = p

        return dummy.next
        


